### PR TITLE
feat(crontask): clean stale queued notifications

### PR DESCRIPTION
**Pull request by Rom1-B**
Added a job to trash browser notifications not sent for X days.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !25082

replace https://github.com/glpi-project/glpi/pull/13070

### CODE REVIEW

File: src/QueuedNotification.php [hunk 1]
```diff
@@ -500,6 +500,11 @@ public static function cronInfo($name)
                 return ['description' => __('Clean notification queue'),
                     'parameter'   => __('Days to keep sent emails')
                 ];
+
+            case 'queuednotificationcleanunset':
```
Comment (cedric-anne) [inline src/QueuedNotification.php:504]: 1. "unset" had a typo, and "stale" is probably more appropriate.
2. "ajax" term should be added to be precise enough.
```suggestion
            case 'queuednotificationcleanstaleajax':
```

File: src/QueuedNotification.php [hunk 2]
```diff
@@ -500,6 +500,11 @@ public static function cronInfo($name)
...
+                return ['description' => __('Clean unsent notification queue'),
```
Comment (cedric-anne) [inline src/QueuedNotification.php:505]:
```suggestion
                return ['description' => __('Clean stale queued browser notifications'),
```

File: src/QueuedNotification.php [hunk 3]
```diff
@@ -644,6 +649,42 @@ public static function cronQueuedNotificationClean($task = null)
...
+     *
+     * @return integer either 0 or 1
+     **/
+    public static function cronQueuedNotificationCleanUnset($task = null)
+    {
+        global $DB;
+
+        $vol = 0;
+
+       // Expire mails in queue
```
Comment (cedric-anne) [inline src/QueuedNotification.php:665]:
```suggestion
        // Stale ajax notifications in queue
```

File: src/QueuedNotification.php [hunk 4]
```diff
@@ -644,6 +649,42 @@ public static function cronQueuedNotificationClean($task = null)
...
+        if ($task->fields['param'] > 0) {
+            $secs      = $task->fields['param'] * DAY_TIMESTAMP;
+            $send_time = date("U") - $secs;
```
Comment (cedric-anne) [inline src/QueuedNotification.php:668]: Should be computed on MySQL server side, to prevent timezone offset issues.

File: install/empty_data.php [hunk 5]
```diff
@@ -789,7 +789,17 @@ public function getEmptyData(): array
                 'mode' => CronTask::MODE_INTERNAL,
                 'lastrun' => null,
                 'logs_lifetime' => 30,
-            ]
+            ], [
+                'id' => 45,
+                'itemtype' => 'QueuedNotification',
+                'name' => 'queuednotificationcleanunset',
+                'frequency' => DAY_TIMESTAMP,
+                'param' => null,
```
Comment (cedric-anne) [inline install/empty_data.php:797]: Should this param have a default value?

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Comment by cedric-anne** [comment]
We often have complains about this, so maybe it should be activated by default. In this case, to ensure that administrator is aware of this cleaning, we could:
 - move the parameter in browser notifications config form (/front/notificationajaxsetting.form.php), using, with a "Validity period of notifications" label and a tooltip to explain that notifications are not sent if the validity period expires and that cleaning is made by the cron task;
 - add a where clause to exclude stale notifications in `NotificationAjax::getMyNotifications()`.

File: install/migrations/update_10.0.x_to_10.1.0/queued_notification.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ * @var DB $DB
+ * @var Migration $migration
+ */
+
+// Register crontask
+CronTask::register('QueuedNotification', 'queuednotificationcleanstaleajax', DAY_TIMESTAMP, [
+    'state'     => CronTask::STATE_DISABLE,
+    'mode'      => CronTask::MODE_INTERNAL,
+]);
+
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/queued_notification.php:46]:
```suggestion

$current_config = Config::getConfigurationValues('core');
```

File: src/NotificationAjaxSetting.php [hunk 7]
```diff
@@ -104,6 +104,23 @@ public function showFormConfig()
                     $CFG_GLPI["notifications_ajax_icon_url"] . "' " .
                     "placeholder='{$CFG_GLPI['root_doc']}/pics/glpi.png'/>";
             echo "</td></tr>";
+
+            $crontask = new CronTask();
+            $crontask->getFromDBbyName('QueuedNotification', 'queuednotificationcleanstaleajax');
+            $tooltip = sprintf(
+                __('%1$s: %2$s'),
+                __('Expired notifications will be trashed by the crontask'),
```
Comment (cedric-anne) [inline src/NotificationAjaxSetting.php:112]:
```suggestion
                __('Notifications older than the selected value will not be displayed. Expired notifications will be deleted by the %s crontask.'),
```

File: src/NotificationAjaxSetting.php [hunk 8]
```diff
@@ -104,6 +104,23 @@ public function showFormConfig()
...
+                $crontask->getLink()
+            );
+            echo "<tr class='tab_bg_2'><td>" . __('Validity period of notifications (in days)') .
+              "&nbsp; " . Html::showToolTip($tooltip, ['display' => false]) .
+              "</td>";
+            echo "<td>";
+            Dropdown::showNumber(
+                'notifications_ajax_expiration_delay',
+                ['value' => $CFG_GLPI["notifications_ajax_expiration_delay"]]
```
Comment (cedric-anne) [inline src/NotificationAjaxSetting.php:121]:
```suggestion
                [
                    'toadd' => [0 => __('Unlimited')],
                    'value' => $CFG_GLPI["notifications_ajax_expiration_delay"],
                ]
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]
OK for me.

The only question is, what should be the default validity period on update ? @orthagh

**Comment by orthagh** [comment]
30 days should be reasonable.

File: install/migrations/update_10.0.x_to_10.1.0/queued_notification.php [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+$current_config = Config::getConfigurationValues('core');
+if (!isset($current_config['notifications_ajax_expiration_delay'])) {
+    $migration->addConfig([
+        'notifications_ajax_expiration_delay' => '7'
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/queued_notification.php:50]:
```suggestion
        'notifications_ajax_expiration_delay' => '30'
```

**Review by trasher — APPROVED** [review]

**Comment by thobou-gly** [comment]
Hi,
1st, thank you for implementing this feature, your work will be much appreciated, at least by me !
Any insight when 10.1 will be released so we all can enjoy it ?
Any way to vote this feature to be implemented sooner ?

**Pull request merged**